File: frontend/lib/patientFilters.ts
/**
 * The generic, per-column-type filter system for the All Patients table -- the
 * frontend's mirror of the backend's `app.schemas.patient_filters` registry (same field
 * keys, same operator sets per type), so "what can I filter by, and how" is answered
 * identically on both sides rather than risking the two silently drifting apart.
 *
 * Replaces the earlier fixed filter panel (one hardcoded UI field per named backend
 * param -- `source`, `age_min`/`age_max`, etc.) with a small column registry that both
 * the "Add Filter" builder (`PatientFilters`) and column-header click-to-sort
 * (`PatientTable`) read from, so adding a new filterable/sortable column in the future
 * means adding one entry here, not touching UI code in two different places.
 */

export type ColumnType = "text" | "number" | "date" | "enum";

export interface ColumnDef {
  key: string;
  label: string;
  type: ColumnType;
  /** Whether this column supports both header-click sorting and (for non-phone columns) filtering. */
  sortable: boolean;
  /** `enum` columns only: the raw backend value + its display label, for the value picker. */
  enumOptions?: { value: string; label: string }[];
}

/**
 * Every All Patients column, in table-display order. `phone` is deliberately marked
 * `sortable: false` and is never filterable at all (see `FILTERABLE_COLUMNS` below) --
 * per direct product decision, a phone number is an identifier, not a value with a
 * meaningful order or a small set of comparable states, the same reasoning the backend's
 * `FIELD_TYPES` registry excludes it for.
 */
export const PATIENT_COLUMNS: ColumnDef[] = [
  { key: "name", label: "Name", type: "text", sortable: true },
  { key: "phone", label: "Phone", type: "text", sortable: false },
  { key: "email", label: "Email", type: "text", sortable: true },
  { key: "age", label: "Age", type: "number", sortable: true },
  { key: "gender", label: "Gender", type: "enum", sortable: true, enumOptions: [
    { value: "male", label: "Male" }, { value: "female", label: "Female" }, { value: "other", label: "Other" },
  ] },
  { key: "source", label: "Source", type: "enum", sortable: true, enumOptions: [
    { value: "in_person", label: "In Person" }, { value: "phone", label: "Phone" }, { value: "instagram", label: "Instagram" },
    { value: "tiktok", label: "TikTok" }, { value: "google", label: "Google" }, { value: "website", label: "Website" },
  ] },
  { key: "created_date", label: "Joined", type: "date", sortable: true },
  { key: "appointment_count", label: "Appointments", type: "number", sortable: true },
  { key: "total_spent_cents", label: "Spent", type: "number", sortable: true },
  { key: "last_appointment_date", label: "Last Appointment", type: "date", sortable: true },
];

/** Columns the "Add Filter" builder can offer -- every column except Phone. */
export const FILTERABLE_COLUMNS = PATIENT_COLUMNS.filter((c) => c.key !== "phone");

export function columnByKey(key: string): ColumnDef | undefined {
  return PATIENT_COLUMNS.find((c) => c.key === key);
}

/** One `{field, operator, value(s)}` filter condition -- mirrors the backend's `PatientFilterCondition`. */
export interface PatientFilterCondition {
  field: string;
  operator: string;
  value?: string;
  value2?: string;
  values?: string[];
}

export const OPERATORS_BY_TYPE: Record<ColumnType, { value: string; label: string }[]> = {
  text: [
    { value: "contains", label: "contains" },
    { value: "not_contains", label: "does not contain" },
    { value: "equals", label: "is" },
    { value: "not_equals", label: "is not" },
  ],
  number: [
    { value: "eq", label: "=" },
    { value: "ne", label: "≠" },
    { value: "gt", label: ">" },
    { value: "gte", label: "≥" },
    { value: "lt", label: "<" },
    { value: "lte", label: "≤" },
    { value: "between", label: "between" },
  ],
  date: [
    { value: "on", label: "on" },
    { value: "not_on", label: "not on" },
    { value: "before", label: "before" },
    { value: "after", label: "after" },
    { value: "on_or_before", label: "on or before" },
    { value: "on_or_after", label: "on or after" },
    { value: "between", label: "between" },
  ],
  enum: [
    { value: "is", label: "is" },
    { value: "is_not", label: "is not" },
    { value: "is_any_of", label: "is any of" },
    { value: "is_none_of", label: "is none of" },
  ],
};

const MULTI_VALUE_OPERATORS = new Set(["is_any_of", "is_none_of"]);
const TWO_VALUE_OPERATORS = new Set(["between"]);

/** How a value editor should render for a given operator: one field, two (a range), or a multi-select. */
export function valueShapeFor(operator: string): "single" | "double" | "multi" {
  if (MULTI_VALUE_OPERATORS.has(operator)) return "multi";
  if (TWO_VALUE_OPERATORS.has(operator)) return "double";
  return "single";
}

/**
 * Whether a filter condition has everything its operator needs to be sent to the API --
 * a condition is kept in local UI state the moment a column is picked (so its row
 * renders), but must not be sent to the backend (and would 400 if it were) until its
 * value(s) are actually filled in. `PatientTable` filters the array down to only
 * complete conditions right before calling `api.getPatients`.
 */
export function isCompleteFilterCondition(condition: PatientFilterCondition): boolean {
  if (valueShapeFor(condition.operator) === "multi") return !!condition.values && condition.values.length > 0;
  if (valueShapeFor(condition.operator) === "double") return !!condition.value && !!condition.value2;
  return !!condition.value;
}

/** A fresh, incomplete draft condition for a newly added filter row, defaulting to the first non-phone column. */
export function newDraftCondition(): PatientFilterCondition {
  const column = FILTERABLE_COLUMNS[0];
  return { field: column.key, operator: OPERATORS_BY_TYPE[column.type][0].value };
}

File: backend/app/schemas/patient_filters.py
"""The generic, per-column-type filter representation for the Patient Table page.

Replaces the earlier All Patients filter shape (a fixed set of named query params --
`source`, `gender`, `age_min`/`age_max`, `created_from`/`created_to`,
`min_total_spent_cents`) with a small, general filter condition that scales to any
column without a new named param per field: `{field, operator, value(s)}`. This is also
squarely in the "AI-ready" spirit the spec calls for (a reusable, composable query shape
a future natural-language-query service could construct directly, rather than a
bespoke per-page parameter list).

`FIELD_TYPES` is the single source of truth for which columns exist and what kind of
data each holds; `OPERATORS_BY_TYPE` is the single source of truth for which operators
are valid for each kind. Both the router (validating a request) and a future
natural-language layer would consult these same two dicts, so "what can I filter by, and
how" never has to be answered twice.

`phone` is deliberately absent from `FIELD_TYPES` -- per direct product decision, phone
numbers aren't a meaningful thing to filter OR sort by (they're an identifier, not a
value with a meaningful order or a small set of comparable states), the same reasoning
that excludes it from column-header sorting everywhere else in this app.
"""

from pydantic import BaseModel, model_validator

# Every filterable All Patients column, and what kind of value it holds. `text`: a
# free-form string (contains/equals-style operators). `number`: a numeric value
# (comparison operators, including "between"). `date`: a calendar date (before/after/
# on, including "between"). `enum`: one of a small fixed set of known values
# (is/is-not/is-any-of/is-none-of).
FIELD_TYPES: dict[str, str] = {
    "name": "text",
    "email": "text",
    "age": "number",
    "gender": "enum",
    "source": "enum",
    "created_date": "date",
    "appointment_count": "number",
    "total_spent_cents": "number",
    "last_appointment_date": "date",
}

OPERATORS_BY_TYPE: dict[str, set[str]] = {
    "text": {"contains", "not_contains", "equals", "not_equals"},
    "number": {"eq", "ne", "gt", "gte", "lt", "lte", "between"},
    "date": {"on", "not_on", "before", "after", "on_or_before", "on_or_after", "between"},
    "enum": {"is", "is_not", "is_any_of", "is_none_of"},
}

# The known values for each enum field -- used only to give a clean 400 (rather than a
# query that silently matches nothing) if a caller sends a value outside this set.
ENUM_VALUES: dict[str, set[str]] = {
    "gender": {"male", "female", "other"},
    "source": {"in_person", "phone", "instagram", "tiktok", "google", "website"},
}

_MULTI_VALUE_OPERATORS = {"is_any_of", "is_none_of"}
_TWO_VALUE_OPERATORS = {"between"}


class PatientFilterCondition(BaseModel):
    """One `{field, operator, value}` filter condition. `value`/`value2` are single
    values (as strings -- callers send numbers/dates as their string representation;
    `app.repositories.patients` parses them back per the field's actual type); `value2`
    is the second bound for `operator="between"`. `values` is used instead of `value`
    for `is_any_of`/`is_none_of`, which compare against a set rather than one value.

    Validated structurally here (right shape of value(s) for the given operator);
    `app.repositories.patients._validate_filter_condition` does the second pass that
    needs the field registry (unknown field, operator not valid for that field's type),
    since that's shared with the actual query-building code and shouldn't be answered
    two different ways in two different places.
    """

    field: str
    operator: str
    value: str | None = None
    value2: str | None = None
    values: list[str] | None = None

    @model_validator(mode="after")
    def _check_value_shape(self) -> "PatientFilterCondition":
        if self.operator in _MULTI_VALUE_OPERATORS:
            if not self.values:
                raise ValueError(f"operator {self.operator!r} requires a non-empty 'values' list")
        elif self.operator in _TWO_VALUE_OPERATORS:
            if self.value is None or self.value2 is None:
                raise ValueError(f"operator {self.operator!r} requires both 'value' and 'value2'")
        elif self.value is None:
            raise ValueError(f"operator {self.operator!r} requires 'value'")
        return self
